File: script/compile.py
# 1. Import solcx
import solcx

# # 2. If you haven't already installed the Solidity compiler, uncomment the following line
# # solcx.install_solc()

# # 3. Compile contract
# # temp_file = solcx.compile_files('../smart-contracts/Incrementer.sol')
# temp_file = solcx.compile_files(
#     "/home/user/ml-on-blockchain/smart-contracts/incrementer.sol"
# )

# # 4. Export contract data
# # abi = temp_file['Incrementer.sol:Incrementer']['abi']
# abi = temp_file[
#     "/home/user/ml-on-blockchain/smart-contracts/incrementer.sol:Incrementer"
# ]["abi"]
# bytecode = temp_file[
#     "/home/user/ml-on-blockchain/smart-contracts/incrementer.sol:Incrementer"
# ]["bin"]


def compile_smart_contract(contract_name):
    """Compile smart contract using solcx.

    Args:
        contract_name (str): Name of the contract to compile.

    Returns:
        dict: Compiled smart contract.
    """
    temp_file = solcx.compile_files(
        f"/home/user/ml-on-blockchain/smart-contracts/{contract_name}.sol"
    )


def get_abi(contract_name):
    """Get ABI of smart contract.
    Require to compile smart contract first.

    Args:
        contract_name (str): Name of the contract to compile.

    Returns:
        dict: ABI of smart contract.
    """
    return temp_file[
        f"/home/user/ml-on-blockchain/smart-contracts/{contract_name}.sol:{contract_name}"
    ]["abi"]


def get_bytecode(contract_name):
    """Get bytecode of smart contract.
    Require to compile smart contract first.

    Args:
        contract_name (str): Name of the contract to compile.

    Returns:
        dict: Bytecode of smart contract.
    """
    return temp_file[
        f"/home/user/ml-on-blockchain/smart-contracts/{contract_name}.sol:{contract_name}"
    ]["bin"]

File: script/compile.js
const fs = require('fs').promises;
//const fs = import('fs.promises');
//import {default as fsWithCallbacks} from 'fs'
//const fs = fsWithCallbacks.promises
const solc = require('solc');
//const solc = import('solc');

async function main() {
  // Load the contract source code
  //const sourceCode = await fs.readFile('../smart-contract/SimpleStorage.sol', 'utf8');
  const sourceCode = await fs.readFile('/home/user/ml-blockchain/simple-transactions/smart-contract/SimpleStorage.sol', 'utf8');
  // Compile the source code and retrieve the ABI and bytecode
  const { abi, bytecode } = compile(sourceCode, 'SimpleStorage');
  // Store the ABI and bytecode into a JSON file
  const artifact = JSON.stringify({ abi, bytecode }, null, 2);
  await fs.writeFile('/home/user/ml-blockchain/simple-transactions/smart-contract/compiled-smart-contract/SimpleStorage.json', artifact);
}

function compile(sourceCode, contractName) {
  // Create the Solidity Compiler Standard Input and Output JSON
  const input = {
   language: 'Solidity',
   sources: { main: { content: sourceCode } },
   settings: { outputSelection: { '*': { '*': ['abi', 'evm.bytecode'] } } },
  };
  // Parse the compiler output to retrieve the ABI and bytecode
    const output = solc.compile(JSON.stringify(input));
    //console.log("output:" + output);
    const artifact = JSON.parse(output).contracts.main[contractName];
    //console.log("artifact:" + artifact);
  return {
    abi: artifact.abi,
    bytecode: artifact.evm.bytecode.object,
  };
}

main().then(() => process.exit(0));
